### PR TITLE
Optimize the code for httpproxy

### PR DESCRIPTION
**Pull request by johncming**
There are three optimization points：

1. Change buffer channel to synchronize channel
2. Remove unnecessary assignment
3. Optimize the sending process， add some error checking and  fix a potential bug.

### CODE REVIEW

File: proxy/httpproxy/reverse.go [hunk 1]
```diff
@@ -125,32 +125,38 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
...
 		if atomic.LoadInt32(&requestClosed) == 1 {
 			return
 		}
+
+		res, err = p.transport.RoundTrip(proxyreq)
 		if err != nil {
 			reportRequestDropped(clientreq, failedSendingRequest)
 			plog.Printf("failed to direct request to %s: %v", ep.URL.String(), err)
 			ep.Failed()
+			res = nil
```
Comment (vimalk78) [inline proxy/httpproxy/reverse.go]: potentially `reportRequestDropped` will be updated twice

File: proxy/httpproxy/reverse.go [hunk 2]
```diff
@@ -107,7 +107,7 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 	}
 
 	var requestClosed int32
-	completeCh := make(chan bool, 1)
+	completeCh := make(chan struct{})
```
Comment (vimalk78) [inline proxy/httpproxy/reverse.go]: looks good

File: proxy/httpproxy/reverse.go [hunk 3]
```diff
@@ -168,8 +174,14 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 	removeSingleHopHeaders(&res.Header)
 	copyHeader(rw.Header(), res.Header)
 
+	if atomic.LoadInt32(&requestClosed) == 1 {
```
[the remark this answers is not in the document]
Comment (johncming) [reply proxy/httpproxy/reverse.go]: yes. If change synchronization, the change will be big.

File: proxy/httpproxy/reverse.go (hunk 1 above)
Comment (johncming) [reply proxy/httpproxy/reverse.go]: This metric corresponds to the number of failures of proxy to backend. If have 5 back ends,  the metric will increase by 5 in the case of failure

**Pull request closed without merging**